deps/chakrashim/src/v8resolver.cc

```diff
@@ -21,164 +21,154 @@
-      return Local<Resolver>();
-    }
+#include <array>
 
-    if (jsrt::SetProperty(
-          resolver,
-          jsrt::CachedPropertyIdRef::promise,
-          promise) != JsNoError) {
-      return Local<Resolver>();
-    }
-
-    if (jsrt::SetProperty(
-          resolver,
-          jsrt::CachedPropertyIdRef::resolve,
-          resolve) != JsNoError) {
-      return Local<Resolver>();
-    }
+namespace v8 {
 
-    if (jsrt::SetProperty(
-          resolver,
-          jsrt::CachedPropertyIdRef::reject,
-          reject) != JsNoError) {
-      return Local<Resolver>();
-    }
+class PromiseResolverData : public ExternalData {
+ public:
+  static const ExternalDataTypes ExternalDataType =
+    ExternalDataTypes::PromiseResolverData;
+
+ private:
+  Persistent<Value> resolve;
+  Persistent<Value> reject;
+
+ public:
+  PromiseResolverData(Local<Value> resolve,
+                      Local<Value> reject)
+      : ExternalData(ExternalDataType),
+        resolve(nullptr, resolve),
+        reject(nullptr, reject) {}
+
+  ~PromiseResolverData() {
+    resolve.Reset();
+    reject.Reset();
+  }
 
-    JsValueRef state;
-    jsrt::UintToValue(0, &state);
-    if (jsrt::SetProperty(
-          promise,
-          jsrt::CachedPropertyIdRef::value,
-          state) != JsNoError) {
-      return Local<Resolver>();
+  static void CHAKRA_CALLBACK FinalizeCallback(void *data) {
+    if (data != nullptr) {
+      PromiseResolverData* promiseResolverData =
+        reinterpret_cast<PromiseResolverData*>(data);
+      delete promiseResolverData;
     }
+  }
 
-    return Local<Resolver>::New(static_cast<Resolver*>(resolver));
+  JsErrorCode Resolve(Local<Value> value) {
+    JsValueRef result = nullptr;
+    std::array<JsValueRef, 2> args = { jsrt::GetUndefined(), *value };
+    return JsCallFunction(*resolve, args.data(), args.size(), &result);
   }
 
-  Local<Resolver> Resolver::New(Isolate* isolate) {
-    return New(isolate->GetCurrentContext()).ToLocalChecked();
+  JsErrorCode Reject(Local<Value> value) {
+    JsValueRef result = nullptr;
+    std::array<JsValueRef, 2> args = { jsrt::GetUndefined(), *value };
+    return JsCallFunction(*reject, args.data(), args.size(), &result);
   }
+};
 
-  Local<Promise> Resolver::GetPromise() {
-    JsValueRef promise;
-    if (jsrt::GetProperty(
-          this,
-          jsrt::CachedPropertyIdRef::promise,
-          &promise) != JsNoError) {
-      return Local<Promise>();
-    }
-    return Local<Promise>::New(static_cast<Promise*>(promise));
+MaybeLocal<Promise::Resolver> Promise::Resolver::New(Local<Context> context) {
+  JsValueRef promise, resolve, reject;
+  if (JsCreatePromise(&promise, &resolve, &reject) != JsNoError) {
+    return Local<Promise::Resolver>();
   }
 
-  Resolver* Resolver::Cast(Value* obj) {
-    return static_cast<Resolver*>(obj);
+  PromiseResolverData* data = new PromiseResolverData(resolve, reject);
+  if (jsrt::AddExternalData(
```

Comment:
Are chakrashim's news throwing? If not, need a check here that data is not null

Reply:
Good question, I believe they are, but I'm not sure.  Looking at the other instances of `new` I don't see any null checks, so if it's no-throw then we'll have to clean them all up.